rskj-core/src/main/java/co/rsk/peg/PeginInformation.java

```diff
@@ -0,0 +1,131 @@
+import org.ethereum.config.blockchain.upgrades.ConsensusRule;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PeginInformation {
+
+    private static final Logger logger = LoggerFactory.getLogger(PeginInformation.class);
+
+    private final BtcLockSenderProvider btcLockSenderProvider;
+    private final PeginInstructionsProvider peginInstructionsProvider;
+    private final ActivationConfig.ForBlock activations;
+
+    private int protocolVersion;
+    private RskAddress rskDestinationAddress;
+    private Address btcRefundAddress;
+    private Address senderBtcAddress;
+    private TxSenderAddressType senderBtcAddressType;
+
+    public PeginInformation(
+        BtcLockSenderProvider btcLockSenderProvider,
+        PeginInstructionsProvider peginInstructionsProvider,
+        ActivationConfig.ForBlock activations) {
+        this.btcLockSenderProvider = btcLockSenderProvider;
+        this.peginInstructionsProvider = peginInstructionsProvider;
+        this.activations = activations;
+        this.protocolVersion = -1; // Set an invalid value by default
+        this.senderBtcAddressType = TxSenderAddressType.UNKNOWN;
+    }
+
+    public int getProtocolVersion() {
+        return protocolVersion;
+    }
+
+    public RskAddress getRskDestinationAddress() {
+        return this.rskDestinationAddress;
+    }
+
+    public Address getBtcRefundAddress() {
+        return this.btcRefundAddress;
+    }
+
+    public Address getSenderBtcAddress() {
+        return this.senderBtcAddress;
+    }
+
+    public TxSenderAddressType getSenderBtcAddressType() {
+        return this.senderBtcAddressType;
+    }
+
+    public void parse(BtcTransaction btcTx) throws PeginInstructionsException {
+        logger.trace("[parse] Trying to parse peg-in information from btc tx {}", btcTx.getHash());
+
+        // Get information from tx sender first
+        Optional<BtcLockSender> btcLockSenderOptional = btcLockSenderProvider.tryGetBtcLockSender(btcTx);
+        if (btcLockSenderOptional.isPresent()) {
+            BtcLockSender btcLockSender = btcLockSenderOptional.get();
+            parseFromBtcLockSender(btcLockSender);
+        }
+
+        // If HF is active and peg-in instructions were provided then override the info obtained from BtcLockSender
+        Optional<PeginInstructions> peginInstructionsOptional = Optional.empty();
+        if (activations.isActive(ConsensusRule.RSKIP170)) {
+            peginInstructionsOptional = peginInstructionsProvider.buildPeginInstructions(btcTx);
+            if (peginInstructionsOptional.isPresent()) {
+                PeginInstructions peginInstructions = peginInstructionsOptional.get();
+                parseFromPeginInstructions(peginInstructions);
+            }
+        }
+
+        // If BtcLockSender could not be parsed and peg-in instructions were not provided, then this tx can't be processed
+        if(!btcLockSenderOptional.isPresent() && !peginInstructionsOptional.isPresent()) {
```

Comment:
I could be wrong on this, but I feel that we have an unexpected change in consensus here.

If we send a transaction from a wp2pkh without presenting the witness, before the activation of the rskip170, it used to reject the transaction and send the money back because with no sender the transaction wouldn't be lockeable, but now it does not seem to send the money back. Or, in general, non lockeable transaction because there's no sending, before rskip170 used to generate a rejection release while it does not happen anymore.